packages/block-editor/src/store/actions.js

```diff
@@ -1282,3 +1283,55 @@ export function setHasControlledInnerBlocks(
 		clientId,
 	};
 }
+
+// This should not be here probably, just experimenting.
+
+/**
+ * Returns an action object used in storing tools panel states for specific blocks.
+ *
+ * @param {string}  blockName Name of the block.
+ * @param {?string} panelName Name of the ToolsPanel/block support
+ */
+
+export const setToolsPanelState = (
+	blockName,
+	panelName,
+	panelItemState
+) => ( { registry } ) => {
+	const existingState =
+		registry
+			.select( preferencesStore )
+			.get( 'core/block-inspector', 'toolsPanel' ) ?? {};
+
+	registry
+		.dispatch( preferencesStore )
+		.set( 'core/block-inspector', 'toolsPanel', {
+			...existingState,
+			[ blockName ]: {
+				...existingState?.[ blockName ],
+				[ panelName ]: {
+					...existingState?.[ blockName ]?.[ panelName ],
+					...panelItemState,
+				},
+			},
+		} );
+};
```

Comment:
It's great to see new usage of the preferences store. 🎉 

It'd be interesting to try to building a react hook instead of using selectors or actions. A lot of the existing preferences use selector or actions only because of backwards compatibility, but I think hooks will be a nicer interface for new things:
```js
const [ isToolsPanelEnabled, setIsToolsPanelEnabled, resetToolsPanel ] = useToolsPanelState( blockName, panelName );
```

Reply:
Great idea. 🎉 That'd be something I'd definitely implement if this experiment grows legs.

I slapped this PR together to test the appetite for saving ToolsPanels preferences, but after having chatted with @aaronrobertshaw, it might be a bit premature, or at least better to wait until the ToolsPanel UX and UI have stabilized.

Reply:
I'm not completely opposed to storing the preferences and I'd say the `ToolsPanel` UI is stable enough for this if we choose to proceed.

The concern was more that in previous discussions around doing this, we had some people prefer the current UX rather than blocks remembering what was previously toggled on.

A few options open to us are:
1. Leave it as is. 
    - You only turn on what you want for any given block instance.
    - Anything you have already customized will be shown along with defaults next time the instance is selected.
    - Optional controls that need toggling on were supposed to be the less commonly used options, if there is something most users are having to toggle on for every block, we can make it a default control.
2. Remember what's been toggled on per block type (as per this PR)
    - Has pros and cons as well
    - If you have multiple block instances that you wish to configure the same control with different values it could save some clicks.
    - If configuring multiple instances with the same value that sounds better suited to Global Styles
    - If only configuring one or two blocks out of several, it would be restoring the sidebar clutter that the `ToolsPanel` paradigm was intending to reduce.
3. Allow per-user opt in to option 2

The bigger issue people have with the `ToolsPanel` doesn't appear so much to be the "remembering of what's toggled on" but the obfuscation of how to reset controls.

It would be great to get @jasmussen and other design folks' opinions on this and how it might affect our goals for cleaning up the sidebar.


---

In terms of the selectors/actions vs custom hooks, I like the hooks approach 👍

Reply:
Thank you for the ping! Just to be sure I read this right, if you add margin to one group block, then it's also visible on other group blocks from that moment onward?

If so, I'd definitely hold off on this, as it goes against the problem the ToolsPanel was meant to solve. A good way to think of the ToolsPanel is that it's meant to manage visibility of _visual changes_, not managing UI. I.e. if you add a margin to a group block, you make a visual change to _that selected group_. This is important for global styles as well, in terms of which defaults we carefully curate on a per-block basis, and which controls a theme might want to intentionally disable.

The calls for making easier to understand how the clearing/resetting of toolspanel controls works are loud and clear, and there are definitely opportunities in improving that interface, whether with subheadings, clearer labels, clearer icons, and what else we can think of. Let's see what we can do there. 

CC: @mtias

Reply:
Thanks for chiming in folks. This was just the type of feedback I was after.

> if you add margin to one group block, then it's also visible on other group blocks from that moment onward?

That was the general idea, albeit this was just a PR to discuss the issue. 

> If so, I'd definitely hold off on this, as it goes against the problem the ToolsPanel was meant to solve. A good way to think of the ToolsPanel is that it's meant to manage visibility of visual changes, not managing UI. 

Good to know, I appreciate the explanation and the feedback! The idea wasn't so much for the ToolsPanel to manage the UI, rather the UI match the preferences of the user. Ultimately I saw it as a general feature users could toggle on and off.

But yeah, I can see how it goes against the general principle of the visible controls reflecting the state of the block's style, and agree we should hold off until things mature and we've gathered/acted upon all the available feedback.

Thanks again! 👍 

I'll just close this PR for now. If the issue every arises at least we have a record of the pros and cons.

Reply:
At some point we could possibly revisit some user configuration of persistently visible overrides. If we did that, we could consider a UI for it that lived in global styles, such as a "pinning" button if you dived into all blocks > block of your choice > tools panel > ellipsis menu > pin next to the item you always want. Something to keep in mind if the use case keeps coming up!

Reply:
One slightly frustrating issue that I discovered is if you do the following:
1. Select a block
2. Select a tool from the drop down
3. Click another block (which a user might do if they wanted to copy/paste a value from that block)
4. Go back to the first selected block, observe the tool you just selected is now not visible and needs to be selected again.

It is fairly minor, but it's something that just doesn't feel right when using the feature.